### ISSUE TITLE
Example in [allocator.requirements.general] incorrectly uses launder?

### ISSUE THREAD

**Issue by languagelawyer** [cplusplus/draft#4553]
From https://stackoverflow.com/q/66755218

[[tab:cpp17.allocator]](https://eel.is/c++draft/tab:cpp17.allocator) has the following [example](https://eel.is/c++draft/tab:cpp17.allocator#row-18-column-3-example-1) [added by p0593r6](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0593r6.html#16535-table-34-cpp17allocator-requirements-tabcpp17allocator):
> When reusing storage denoted by some pointer value `p`, `launder(reinterpret_­cast<T*>(new (p) byte[n * sizeof(T)]))` can be used to implicitly create a suitable array object and obtain a pointer to it.

If `T` is not an implicit-lifetime type, then, even if an object of  type `T[m ≤ n]` is created, the lifetimes of the array elements are not started, which means `launder`'s preconditions are violated:
> _Preconditions_: `p` represents the address `A` of a byte in memory. An object `X` **that is within its lifetime** and whose type is similar to `T` is located at the address `A`.

(If `T` is an implicit-lifetime type, or even for all types `T`, `launder` may not even be required, see below)

@zygoloid As it is shown in the SO question, one can't obtain a pointer to array element of non-implicit-lifetime type `T` by just casting the allocating operation to `T*`. This depends on the interpretation of _These operations select one of the implicitly-created objects ..._: does it refer to only «top-level» objects of an implicit-lifetime types, or subobjects such as array elements or member subobjects can also be selected by an operation creating an array or class object?

**Comment by theundergroundsorcerer** [comment]
The usage is indeed incorrect. The valid cast is `launder(reinterpret_­cast<T(*)[n]>(new (p) byte[n * sizeof(T)]))` (May be `n` is not necessary) I believe. The problem is that value of the array (as a pointer) is valid pointer for purposes of arithmetic and casting to it and laundering it should be allowed as long as I do not de-reference it, (without constructing an object first).

**Comment by xmh0511** [comment]
> These operations select one of the implicitly-created objects **whose address is the address of the start of the region of storage**, and produce a pointer value that points to that object, if that value would result in the program having defined behavior.  

Incidentally, the standard never specifies which implicitly created objects whose addresses are the address of the start of the region of storage.

**Comment by jwakely** [comment]
That's intentional. The right one is selected, by magic, if that would make the program correct.

None of this is editorial.

**Comment by xmh0511** [comment]
> That's intentional. The right one is selected, by magic, if that would make the program correct.

So, we cannot plainly determine whether a program is well-formed. For instance
````cpp
struct X{
   int x;
};
auto ptr = (X*)malloc(sizeof(X));
ptr->x = 0;// is well-formed?
````
No one can say the return value of `malloc` can be the address of object `X`. Although `X` is implicitly-lifetime type, however, we don't know whether that object's address is the address of the start of the region of storage.

**Comment by jwakely** [comment]
No, that's not what the wording means. There is no "implementor of the function `malloc`," because it's part of the implementation and so behaves as the standard says (not as some developer says). The standard says that if beginning the lifetime of an object of type `X` at that location would make the program have defined behaviour, then that's what happens. So the answer to the "is well-formed?" question is yes. By definition. If creating an `X` there makes it well-formed, then that's what happens, and so it's well-formed.

**Comment by languagelawyer** [comment]
«well-formed» should not be confused with «well-defined»

**Comment by jwakely** [comment]
Yes, and `ptr->x = 0` is always well-formed when `ptr` has type `X*`. It's also well-defined because `malloc` implicitly creates objects of the type needed to make it have defined behaviour.

**Comment by xmh0511** [comment]
@jwakely However, the standard/implementation never says that the object of type `X` would be located at the start address of that region. This is the confusion here.

**Comment by xmh0511** [comment]
Cite the quotes of the definition of `malloc` in C
> Description: 
>> - The malloc function allocates space for an object whose size is specified by size and
whose value is indeterminate.
> 
> Returns
>> - The malloc function returns either a null pointer or **a pointer to the allocated space**.  

````cpp
struct T{
   int i;
   char c;
};
auto tptr = (T*)malloc(sizeof(T));
tptr-> a = 0;
auto iptr = (int*)malloc(sizeof(T));
*iptr = 0;
````
The standard only specifies that an object of type `T` and the member subobject `T::i` is pointer-interconvertible, hence they have the same address. However, neither in C++ nor C standard ever specifies that the addresses of a type `T` object and its subobject `T::i` are identical to the address of the start of the region of the storage that is allocated by the `malloc` function. we say the above example is well-defined is just based on the assumption that these objects **whose addresses are the address of the start of the allocated space**. However, this assumption has no formal definition to prove in the current standard.

**Comment by xmh0511** [comment]
@languagelawyer 
> the lifetimes of the array elements are not started

The lifetimes of the array elements are started, as per [expr.new] p13
> A new-expression that creates an object of type T initializes that object as follows:
>> - If the new-initializer is omitted, the object is default-initialized ([dcl.init]).

Since the object is an array, default-initialize an array object has an equivalent effect with that as per [dcl.init.general] p7
> To default-initialize an object of type T means:
>> - [...]
>> - If T is an array type, each element is default-initialized.

Thus,  each element of the array object is default-initialized, which obeys [basic.life] p1
> The lifetime of an object of type T begins when:
>> - storage with the proper alignment and size for type T is obtained, and
>> - its initialization (if any) is complete (**including vacuous initialization**) ([dcl.init]),

**Comment by frederick-vs-ja** [comment]
[P2590R0](https://wg21.link/p2590r0) addressed this issue by replace the use with `start_lifetime_as_array<T>(p, n)`, but this change was removed in later revisions.

**Comment by jensmaurer** [comment]
Yes, because that's a situation where we're not intending to re-use the existing bytes for the to-be object representation. As-is, the optimizer can apply dead-store elimination for any store to the affected region.

**Comment by frederick-vs-ja** [comment]
`start_lifetime_as_array<T>(new (p) byte[n * sizeof(T)], n)` seemingly works as intended. Although there's an open issue CWG1997 indicating that it may be unclear whether the new-expression renders the contents in the storage indeterminate.

**Comment by jensmaurer** [comment]
That core issue far predates `std::start_lifetime_as` and the updated understanding of what placement-new should do in this case.

**Comment by languagelawyer** [comment]
The goal of CWG1997, in the first place, should be to clarify that the lifetime of objects created by new-expression using non-allocating `operator new` starts, because «obtaining storage» is a precondition to lifetime start, and somehow there is doubt that non-allocating `operator new` obtains storage 🤡

**Comment by MX20** [comment]
Although std::launder  should not be required at all and everything should correct itself.

1) According to the standard, "operations" that begin the lifetime of std::byte array should implicitly create an object! http://eel.is/c++draft/basic.memobj#intro.object-13
2) And in the case of non-implicit type, it would be just a pointer * before * the start of a lifetime!

Example for (1). If the object is created implicitly, then placement new shall return a pointer to it. Therefore this is a perfectly valid code point to the first element of the array. 

    struct A { int p; };
    std::byte s[100 * sizeof(A)];
    reinterpret_cast<A*>(new (s) std::byte[100 * sizeof(A)])[4].p = 100; //OK?

**Comment by languagelawyer** [comment]
@MX20 `new ... T[N]`, for  N > 0, always returns the pointer to the first element of the «array of N `T`» http://eel.is/c++draft/expr.new#10.sentence-1

**Comment by MX20** [comment]
Right , but what does it change? For *non* implicitly created types it may be the pointer to the first element before its lifetime started which can be used in limited ways  - just enough to create an object in it using another placement new

**Comment by languagelawyer** [comment]
> Right , but what does it change?

That `new (s) std::byte[…]` always returns a pointer to the first element of the array of some `std::byte`s, and implicitly created objects do not affect this.

**Comment by MX20** [comment]
I see , fair enough

**Comment by zygoloid** [comment]
I think there's an interesting decision for CWG to make here. I agree with the issue description:

> > When reusing storage denoted by some pointer value `p`, `launder(reinterpret_­cast<T*>(new (p) byte[n * sizeof(T)]))` can be used to implicitly create a suitable array object and obtain a pointer to it.
> 
> If `T` is not an implicit-lifetime type, then, even if an object of type `T[m ≤ n]` is created, the lifetimes of the array elements are not started, which means `launder`'s preconditions are violated:
> 
> > _Preconditions_: `p` represents the address `A` of a byte in memory. An object `X` **that is within its lifetime** and whose type is similar to `T` is located at the address `A`.
> 
> (If `T` is an implicit-lifetime type, or even for all types `T`, `launder` may not even be required, see below)

I think we can resolve the disagreement between the claim about a valid use of `launder` and the precondition of `launder` in (at least) two ways. We could say that:

1) The claim is wrong, and we should update the text to suggest a different approach.
2) The precondition of `launder` is wrong, and it should alternatively suffice for there to exist an array object of type `T[n]` within its lifetime such that `p` represents the address of an element of the array, which need not be within its lifetime.
3) We want a separate `launder_array` function for this case.

Option 2 is probably the simplest if it works.

**Comment by languagelawyer** [comment]
> 2. The precondition of `launder` is wrong, and it should alternatively suffice for there to exist an array object of type `T[n]` within its lifetime such that `p` represents the address of an element of the array, which need not be within its lifetime.
>
> Option 2 is probably the simplest if it works.

```c++
int arr[1];
for (int i = 0; i < INT_MAX; i++) ::new (&arr[0]) int {};
std::launder(&arr[0]); // which of INT_MAX+1 objects std::launder returns pointer to?
```

**Comment by frederick-vs-ja** [comment]
> ```c++
> int arr[1];
> for (int i = 0; i < INT_MAX; i++) ::new (&arr[0]) int {};
> std::launder(&arr[0]); // which of INT_MAX+1 objects std::launder returns pointer to?
> ```

The rule for transparently replacing makes all of these pointer values equivalent, IIUC.

**Comment by frederick-vs-ja** [comment]
LWG4130
